.github/workflows/release.yml

```diff
@@ -4,18 +4,145 @@ on:
+            echo "version=${VERSION:1}" >> "$GITHUB_OUTPUT"
+          else
+            echo "Scope $VERSION does not follow v<version> naming scheme"
+            exit 1
+          fi
+        env:
+          VERSION_OVERRIDE: ${{ inputs.version-override }}
+      - uses: actions/checkout@v4
+      - name: Generate release body
+        run: |
+          sed "s/:VERSION/$VERSION/g" < .github/release-template.md > release.md
+          cat release.md
+        env:
+          VERSION: ${{ steps.getrelease.outputs.version }}
+      - name: Upload release body
+        uses: actions/upload-artifact@v4
+        with:
+          name: release-body
+          path: release.md
+  build-wheels:
+    runs-on: ${{ matrix.os }}
+    needs: [extract-params]
+    strategy:
+      fail-fast: false
+      matrix:
+        os:
+          - ubuntu-latest
+          - ubuntu-24.04-arm
+          - windows-latest
+          - macos-latest
+    steps:
+      - uses: actions/checkout@v4
+      - uses: ./.github/actions/checkout-for-release
+        with:
+          version: ${{ needs.extract-params.outputs.version }}
+      - name: Build wheels
+        uses: pypa/cibuildwheel@v3.0.0
+      - uses: actions/upload-artifact@v4
+        with:
+          name: wheels-${{ matrix.os }}-${{ strategy.job-index }}
+          path: ./wheelhouse/*.whl
+  build-sdist:
+    runs-on: ubuntu-latest
+    needs: [extract-params]
+    steps:
+      - uses: actions/checkout@v4
+      - uses: ./.github/actions/checkout-for-release
+        with:
+          version: ${{ needs.extract-params.outputs.version }}
+      - name: Build source distribution
+        run: uv build --sdist
+      - uses: actions/upload-artifact@v4
+        with:
+          name: sdist
+          path: ./dist/*.tar.gz
+  publish:
+    name: Publish release artifacts
+    needs: [extract-params, build-sdist, build-wheels]
     runs-on: ubuntu-latest
-    environment: release
+    environment: ${{ needs.extract-params.outputs.publish-env }}
     permissions:
+      # we use PyPI's trusted publisher model -> expose identity token
       id-token: write
     steps:
-      - uses: actions/checkout@v4
-      - uses: actions/setup-python@v5
-      - run: pip install build
-      - run: python -m build
-      - name: Publish package distributions to PyPI
+      - name: Download wheels
+        uses: actions/download-artifact@v4
+        with:
+          pattern: wheels-*
+          path: dist/
+          merge-multiple: 'true'
+      - name: Download source distribution
+        uses: actions/download-artifact@v4
+        with:
+          name: sdist
+          path: dist/
+      - name: Download release body
+        uses: actions/download-artifact@v4
+        with:
+          name: release-body
+          path: release-body
+      - name: Upload to PyPI
         uses: pypa/gh-action-pypi-publish@release/v1
+        with:
+          repository-url: ${{ vars.REPOSITORY_URL }}
```

Comment:
@kislyuk two things to check:

 - The release environment on the PyPI publisher is called `pypi`, right? (going off the comments in #203)
 - Can you make sure there is a `REPOSITORY_URL` variable defined in that environment with the value `https://upload.pypi.org/legacy/`?
 - While you're at it, if you could create another environment called `test-pypi` with `REPOSITORY_URL` set to `https://test.pypi.org/legacy/`, that'd be nice. Then I can configure the test project on TestPyPI to also accept releases from this repo :).